environment/src/main/java/com/sequenceiq/environment/environment/sync/EnvironmentSyncService.java

```diff
@@ -1,24 +1,103 @@
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.UPGRADE_CCM_FAILED;
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.UPGRADE_CCM_IN_PROGRESS;
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.UPGRADE_CCM_REQUESTED;
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.UPSCALE_FAILED;
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.VERTICAL_SCALE_FAILED;
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status.WAIT_FOR_SYNC;
 
+import java.util.Map;
 import java.util.Optional;
 
 import org.springframework.stereotype.Component;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.sequenceiq.environment.environment.EnvironmentStatus;
 import com.sequenceiq.environment.environment.domain.Environment;
 import com.sequenceiq.environment.environment.service.freeipa.FreeIpaService;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status;
 import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.describe.DescribeFreeIpaResponse;
 
 @Component
 public class EnvironmentSyncService {
 
+    private static final Map<Status, EnvironmentStatus> FREEIPA_STATUS_TO_ENV_STATUS_MAP = Map.ofEntries(
+            Map.entry(REQUESTED, FREEIPA_CREATION_IN_PROGRESS),
+            Map.entry(CREATE_IN_PROGRESS, FREEIPA_CREATION_IN_PROGRESS),
+            Map.entry(AVAILABLE, EnvironmentStatus.AVAILABLE),
+            Map.entry(STACK_AVAILABLE, FREEIPA_CREATION_IN_PROGRESS),
+            Map.entry(DIAGNOSTICS_COLLECTION_IN_PROGRESS, EnvironmentStatus.AVAILABLE),
+            Map.entry(UPDATE_IN_PROGRESS, UPDATE_INITIATED),
+            Map.entry(UPDATE_REQUESTED, UPDATE_INITIATED),
+            Map.entry(UPDATE_FAILED, EnvironmentStatus.UPDATE_FAILED),
+            Map.entry(UPSCALE_FAILED, EnvironmentStatus.AVAILABLE),
+            Map.entry(DOWNSCALE_FAILED, EnvironmentStatus.AVAILABLE),
+            Map.entry(VERTICAL_SCALE_FAILED, VERTICAL_SCALE_ON_FREEIPA_FAILED),
+            Map.entry(REPAIR_FAILED, FREEIPA_UNHEALTHY),
+            Map.entry(CREATE_FAILED, EnvironmentStatus.CREATE_FAILED),
+            Map.entry(DELETE_IN_PROGRESS, FREEIPA_DELETE_IN_PROGRESS),
+            Map.entry(DELETE_FAILED, EnvironmentStatus.DELETE_FAILED),
+            Map.entry(DELETE_COMPLETED, FREEIPA_DELETE_IN_PROGRESS),
+            Map.entry(STOPPED, ENV_STOPPED),
+            Map.entry(STOP_REQUESTED, STOP_FREEIPA_STARTED),
+            Map.entry(START_REQUESTED, START_FREEIPA_STARTED),
+            Map.entry(STOP_IN_PROGRESS, STOP_FREEIPA_STARTED),
+            Map.entry(START_IN_PROGRESS, START_FREEIPA_STARTED),
+            Map.entry(START_FAILED, START_FREEIPA_FAILED),
+            Map.entry(STOP_FAILED, STOP_FREEIPA_FAILED),
+            Map.entry(WAIT_FOR_SYNC, START_SYNCHRONIZE_USERS_STARTED),
+            Map.entry(MAINTENANCE_MODE_ENABLED, EnvironmentStatus.AVAILABLE),
```

Comment:
Is the FreeIPA indeed healthy and ready for operations in this case?

Reply:
most likely :)